backend/src/router/utils/search.ts

```diff
@@ -0,0 +1,114 @@
+    return;
+  } else if (typeof blockId === "string") {
+    selection = selection.where("block_hash", "=", blockId);
+  } else {
+    selection = selection.where("block_height", "=", blockId.toString());
+  }
+  const block = await selection.limit(1).executeTakeFirst();
+  if (block) {
+    return { blockHash: block.hash };
+  }
+};
+
+const getTransactionHash = async (
+  transactionHash: z.infer<typeof validators.transactionHash>
+) => {
+  const selection = await indexerDatabase
+    .selectFrom("transactions")
+    .select("transaction_hash as hash")
+    .where("transaction_hash", "=", transactionHash)
+    .limit(1)
+    .executeTakeFirst();
+  if (selection) {
+    return { transactionHash: selection.hash };
+  }
+};
+
+const getAccountId = async (input: z.infer<typeof validators.accountId>) => {
+  const selection = await indexerDatabase
+    .selectFrom("accounts")
+    .where("account_id", "=", input)
+    .select(["account_id as accountId"])
+    .limit(1)
+    .executeTakeFirst();
+  if (selection) {
+    return { accountId: selection.accountId };
+  }
+};
+
+const getReceiptId = async (
+  receiptId: z.infer<typeof validators.receiptId>
+) => {
+  const selection = await indexerDatabase
+    .selectFrom("receipts")
+    .select([
+      "receipt_id as receiptId",
+      "originated_from_transaction_hash as transactionHash",
+    ])
+    .where("receipt_id", "=", receiptId)
+    .limit(1)
+    .executeTakeFirst();
+  if (selection) {
+    return {
+      transactionHash: selection.transactionHash,
+      receiptId: selection.receiptId,
+    };
+  }
+};
+
+const validateAndFetch = async <T, V extends ZodType<any, any, any>>(
+  input: string,
+  validator: V,
+  fetcher: (input: V["_type"]) => Promise<T>
+): Promise<T | undefined> => {
+  const result = validator.safeParse(input);
+  if (!result.success) {
+    throw new Error("Validation error");
+  }
+  const fetchedResult = await fetcher(result.data);
+  if (!fetchedResult) {
+    throw new Error("Fetch error");
+  }
+  return fetchedResult;
+};
+
+export const router = trpc.router<Context>().query("search", {
+  input: z.strictObject({
+    value: z.string(),
+  }),
+  resolve: async ({ input: { value } }) => {
+    const result = await Promise.any([
+      validateAndFetch(value, blockInput, getBlockHash),
+      validateAndFetch(value, validators.transactionHash, getTransactionHash),
+      validateAndFetch(value, validators.accountId, getAccountId),
+      validateAndFetch(value, validators.receiptId, getReceiptId),
+    ]);
+    if (!result) {
+      return;
+    } else if ("blockHash" in result) {
+      return "/blocks/" + result.blockHash;
+    } else if ("receiptId" in result) {
+      return `/transactions/${result.transactionHash}#${result.receiptId}`;
+    } else if ("transactionHash" in result) {
+      return "/transactions/" + result.transactionHash;
+    } else if ("accountId" in result) {
+      return "/accounts/" + result.accountId;
+    }
```

Comment:
nit: I think it is not a concern of the backend to decide on the redirection URL for the frontend. I would consider returning an object like `{"kind": "block", "id": result.blockHash}`.

Reply:
That's true, I'll change that.
Does that create attack surface? I was thinking about that but didn't find one.

Reply:
@luixo I don't see an attack surface, but it creates unnecessary coupling between the backend and frontend. It is not critical for Explorer, yet it makes sense